jsonrpc2.go

```diff
@@ -76,6 +89,8 @@ func (r *Request) UnmarshalJSON(data []byte) error {
 		Meta   *json.RawMessage `json:"meta,omitempty"`
 		ID     *ID              `json:"id"`
 	}
+	// This is used to get the extra fields, which are not type-safe.
+	r3 := make(map[string]*json.RawMessage)
```

Comment:
re: https://github.com/sourcegraph/jsonrpc2/issues/49#issuecomment-887275984

things i tried:

* make `r2` a `map[string]interface{}`, and it's possible to extract the method, params, meta, and id, _but_ then i needed to marshal all the extra fields into a `*json.RawMessage`.
* make `r2` a `map[string]*json.RawMessage`, and it's now possible to extract the extra fields, but the method, params, meta, and id now need an extra unmarshal.
* make the `RequestField.Value` an `interface{}` so that the first option is easier (but didn't really try that because it felt inconsistent with the `Meta` and `Params` fields). it also defers the marshaling error to the `MarshalJSON`, but that might not be such a big deal since both would happen at the `.Call()` API level.

Reply:
> * make `r2` a `map[string]interface{}`, and it's possible to extract the method, params, meta, and id, _but_ then i needed to marshal all the extra fields into a `*json.RawMessage`.

Can't we just make `RequestField.Value` an `interface{}`? This seems like the easiest option. It does mean if you want to avoid marshalling on extra fields you can't. Is this acceptable?

> * make `r2` a `map[string]*json.RawMessage`, and it's now possible to extract the extra fields, but the method, params, meta, and id now need an extra unmarshal.

This seems like the best option for avoiding the extra work/allocations? It would also simplify our handling of detecting if params is JSON null? I'm a bit cognizant of "doubling" the work this does.

Reply:
> Can't we just make `RequestField.Value` an `interface{}`? This seems like the easiest option. It does mean if you want to avoid marshalling on extra fields you can't. Is this acceptable?

totally acceptable (and in fact will make one particular test easier for us! haha), I just didn't want to start with an option that introduced inconsistency without checking with y'all.

> This seems like the best option for avoiding the extra work/allocations? It would also simplify our handling of detecting if params is JSON null? I'm a bit cognizant of "doubling" the work this does.

it avoids the extra allocations, but introduces extra casts. the amount of work is roughly the same _but_ it makes (our particular usage of) the interface a bit simpler due to the lack of the pre-marshaling for the extra fields.

Reply:
oh wait, i lied: this unmarshaling option also adds extra _marshaling_: the params and meta would now be fully deserialized by the `map[string]interface{}`, so now we would need to marshsal 'em back to a `*json.RawMessage`.